discretisedfield/tests/test_tools.py

```diff
@@ -64,56 +64,96 @@ def test_topological_charge(method):
             getattr(dft, function)(f.sel("z"), method="wrong")
 
 
-def test_emergent_magnetic_field():
-    p1 = (0, 0, 0)
-    p2 = (10, 10, 10)
-    cell = (2, 2, 2)
+@pytest.mark.parametrize("ndim", [1, 2, 3, 4])
+@pytest.mark.parametrize("nvdim", [1, 2, 3, 4])
+def test_emergent_magnetic_field(ndim, nvdim):
+    p1 = (0,) * ndim
+    p2 = (10,) * ndim
+    cell = (2,) * ndim
+    value = (0,) * nvdim
     mesh = df.Mesh(p1=p1, p2=p2, cell=cell)
 
     # f(x, y, z) = (0, 0, 0)
     # -> F(f) = 0
-    f = df.Field(mesh, nvdim=3, value=(0, 0, 0))
+    f = df.Field(mesh, nvdim=nvdim, value=value)
 
-    assert dft.emergent_magnetic_field(f).nvdim == 3
-    assert np.allclose(dft.emergent_magnetic_field(f).mean(), (0, 0, 0))
+    if ndim != 3 or nvdim != 3:
+        with pytest.raises(ValueError):
+            dft.emergent_magnetic_field(f)
+    else:
+        assert dft.emergent_magnetic_field(f).nvdim == 3
+        assert np.allclose(dft.emergent_magnetic_field(f).mean(), (0, 0, 0))
+
+
+@pytest.mark.parametrize("ndim", [1, 2, 3, 4])
+@pytest.mark.parametrize("nvdim", [1, 2, 3, 4])
+def test_neighbouring_cell_angle(ndim, nvdim):
+    p1 = (0,) * ndim
+    p2 = (100,) * ndim
+    n = (10,) * ndim
+    value = (1,) * nvdim
+    region = df.Region(p1=p1, p2=p2, dims=[f"g{i}" for i in range(1, ndim + 1)])
+    mesh = df.Mesh(region=region, n=n)
+    field = df.Field(mesh, nvdim=nvdim, value=value)
+
+    # Exception
+    if nvdim != 3:
+        with pytest.raises(ValueError):
+            dft.neighbouring_cell_angle(field, direction="g1")
+        return
 
     with pytest.raises(ValueError):
-        dft.emergent_magnetic_field(f.x)
-
+        dft.neighbouring_cell_angle(field, direction="x")
 
-def test_neigbouring_cell_angle():
-    p1 = (0, 0, 0)
-    p2 = (100, 100, 100)
-    n = (10, 10, 10)
-    mesh = df.Mesh(p1=p1, p2=p2, n=n)
-    field = df.Field(mesh, nvdim=3, value=(0, 1, 0))
+    with pytest.raises(ValueError):
+        dft.neighbouring_cell_angle(field, direction="g1", units="wrong")
 
-    for direction in "xyz":
+    for direction in field.mesh.region.dims:
         for units in ["rad", "deg"]:
-            sa = dft.neigbouring_cell_angle(field, direction=direction, units=units)
+            sa = dft.neighbouring_cell_angle(field, direction=direction, units=units)
             assert sa.mean() == 0
 
-    # Exceptions
-    scalar_field = df.Field(mesh, nvdim=1, value=5)
-    with pytest.raises(ValueError):
-        dft.neigbouring_cell_angle(scalar_field, direction="x")
-
-    with pytest.raises(ValueError):
-        dft.neigbouring_cell_angle(field, direction="l")
-
-    with pytest.raises(ValueError):
-        dft.neigbouring_cell_angle(field, direction="x", units="wrong")
+    # Check for a value of angle
+    arr_x_coord = mesh.coordinate_field().array[..., 0, np.newaxis].copy()
```

Comment:
```suggestion
    arr_x_coord = mesh.coordinate_field().g1.array
```
Could we do?

Reply:
This unfortunately will ask for the `g1` component of the vector which does not exist. `g1` is a geometric dimension here.

Reply:
Yes, it works but not for `vdims=1` and `vdims=4` 😅 . I have a work-around though, I'll push a commit and tag you again.

Reply:
@swapneelap sorry for the poor explication, I have just committed what I was meaning. Take a look and feel free to remove the commit if you want but it works for all `ndim` and the tests pass but I may be missing something.

Reply:
I had exactly the same thing in mind 😄 . Thank you for the commit!